src/App-custom-step.spec.ts

```diff
@@ -0,0 +1,176 @@
+import 'mocha';
+import sinon from 'sinon';
+import { assert } from 'chai';
+import rewiremock from 'rewiremock';
+import { Override, mergeOverrides } from './test-helpers';
+import {
+  Receiver,
+  ReceiverEvent,
+} from './types';
+import App from './App';
+
+// Fakes
+class FakeReceiver implements Receiver {
+  private bolt: App | undefined;
+
+  public init = (bolt: App) => {
+    this.bolt = bolt;
+  };
+
+  public start = sinon.fake((...params: any[]): Promise<unknown> => Promise.resolve([...params]));
+
+  public stop = sinon.fake((...params: any[]): Promise<unknown> => Promise.resolve([...params]));
+
+  public async sendEvent(event: ReceiverEvent): Promise<void> {
+    return this.bolt?.processEvent(event);
+  }
+}
+
+describe('App function() and function_executed and function-scoped interactivity event processing', () => {
+  let fakeReceiver: FakeReceiver;
+  let dummyAuthorizationResult: { botToken: string; botId: string };
+
+  beforeEach(() => {
+    fakeReceiver = new FakeReceiver();
+    dummyAuthorizationResult = { botToken: '', botId: '' };
+  });
+
+  let app: App;
+
+  beforeEach(async () => {
+    const MockAppNoOverrides = await importApp();
+    app = new MockAppNoOverrides({
+      receiver: fakeReceiver,
+      authorize: sinon.fake.resolves(dummyAuthorizationResult),
+    });
+  });
+
+  it('should add a listener to middleware for each function listener passed to app.function', async () => {
+    /* middleware is a private property on App, so ensure that the step listener
+    is added to the global middleware array. */
+    const { middleware } = (app as any);
+    const noop = () => Promise.resolve();
+
+    assert.equal(middleware.length, 2);
+
+    app.function('callback_id', noop, noop);
+
+    // Bundles two listeners into one middleware, thus 3 here:
+    assert.equal(middleware.length, 3);
+  });
+  it('should enrich custom step context and arguments with function_executed properties like execution ID and inputs', async () => {
+    const noop = () => Promise.resolve();
+    const processStub = sinon.fake.resolves({});
+    const MockAppNoOverrides = await importApp(mergeOverrides(withNoopAppMetadata(), withNoopWebClient(), {
+      './middleware/process': processStub,
+    }));
+    app = new MockAppNoOverrides({
+      receiver: fakeReceiver,
+      authorize: sinon.fake.resolves(dummyAuthorizationResult),
+    });
+    await fakeReceiver.sendEvent({
+      body: {
+        event: {
+          type: 'function_executed',
+          function_execution_id: 'Fx1234',
+          bot_access_token: 'xowfp',
+        },
+        function_data: {
```

Comment:
I had to double-check, but AFAIK, `function_data` only is associated with `block_actions` payloads – not `function_executed`.

```
{
  type: 'function_executed',
  function: {
    id: 'Fn0695V91LGN',
    callback_id: 'sample_function',
    title: 'Sample function 1',
    description: 'Runs sample function',
    type: 'app',
    input_parameters: [ [Object] ],
    output_parameters: [ [Object] ],
    app_id: 'A06A1C8C5FS',
    date_created: 1702069571,
    date_released: 0,
    date_updated: 1723658242,
    date_deleted: 0,
    form_enabled: false
  },
  inputs: { user_id: 'U019GD47LJH' },
  function_execution_id: 'Fx07HMB67Y3S',
  workflow_execution_id: 'Wx07GVLL7ZTP',
  event_ts: '1723658328.841355',
  bot_access_token: 'xwfp-444'
}
```

versus

```
{
  type: 'block_actions',
  team: { id: 'T019KFN8A2W', domain: 'misscoded' },
  enterprise: null,
  user: { id: 'U019GD47LJH', name: 'arenz', team_id: 'T019KFN8A2W' },
  channel: { id: 'D0699LNQ3HT', name: 'directmessage' },
  message: {
    user: 'U069CFDKS03',
    type: 'message',
    ts: '1723658484.749309',
    bot_id: 'B069F2HQSFN',
    app_id: 'A06A1C8C5FS',
    text: 'Click the button to signal the function has completed',
    team: 'T019KFN8A2W',
    blocks: [ [Object] ],
    assistant_app_thread: { title: '', title_blocks: [] }
  },
  container: {
    type: 'message',
    message_ts: '1723658484.749309',
    channel_id: 'D0699LNQ3HT',
    is_ephemeral: false
  },
  actions: [
    {
      block_id: 'I9kjf',
      action_id: 'sample_button',
      type: 'button',
      text: [Object],
      action_ts: '1723658488.667103'
    }
  ],
  api_app_id: 'A06A1C8C5FS',
  state: { values: {} },
  bot_access_token: 'xwfp-444',
  function_data: {
    execution_id: 'Fx07GVMA73NH',
    function: { callback_id: 'sample_function' },
    inputs: { user_id: 'U019GD47LJH' }
  },
  interactivity: {
    interactor: {
      secret: 'xxx',
      id: 'U019GD47LJH'
    },
    interactivity_pointer: '7576612985154.1325532282098.90f3e0899eaced8770aaf84abe894dcd'
  }
}
```